File: TP1/RecursivePalindrome.java
import java.text.Normalizer;

public class RecursivePalindrome{
    static final String inputBreak = "FIM";

    public static boolean isPalindrome(String str, int i, int j) {
        str = Normalizer.normalize(str, Normalizer.Form.NFD);

        if (i >= j) {
            return true;
        }

        if (str.charAt(i) != str.charAt(j)) {
            return false;
        }

        return isPalindrome(str, i + 1, j - 1);
    }

    public static boolean isFim(String str) {
        return str.length() == 3 &&
               str.charAt(0) == 'F' &&
               str.charAt(1) == 'I' &&
               str.charAt(2) == 'M';
    }

    public static void main(String[] args) {
        String str = MyIO.readLine();
        processInput(str);
    }

    public static void processInput(String str) {
        if (isFim(str)) {
            return;
        }

        if (isPalindrome(str, 0, str.length() - 1)) {
            MyIO.println("SIM");
        } else {
            MyIO.println("NAO");
        }

        processInput(MyIO.readLine());
    }
}

File: TP1/RecursivePalindrome.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define TRUE 1
#define FALSE 0
#define INPUT_BREAK "FIM"

bool isPalindrome(char *str, int i, int j) {
    if (i >= j) {
        return TRUE;
    }

    if (str[i] != str[j]) {
        return FALSE;
    }

    return isPalindrome(str, i + 1, j - 1);
}

void processInput() {
    char str[1000];
    fgets(str, sizeof(str), stdin);

    size_t len = strlen(str);
    if (len > 0 && str[len - 1] == '\n') {
        str[len - 1] = '\0';
    }

    if (strcmp(str, INPUT_BREAK) == 0) {
        return;
    }

    if (isPalindrome(str, 0, strlen(str) - 1)) {
        printf("SIM\n");
    } else {
        printf("NAO\n");
    }

    processInput();
}

int main() {
    processInput();
    return 0;
}